DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/fft2d/src/fft2d_demo.cpp

```diff
@@ -0,0 +1,420 @@
+      }
+    }
+    double db = 10 * log(magnitude_sum / noise_sum) / log(10.0);
+
+    std::cout << "Signal to noise ratio on output sample: " << db << std::endl;
+    std::cout << " --> " << (db > 120 ? "PASSED" : "FAILED") << std::endl;
+
+    free(input_data, q);
+    free(output_data, q);
+    free(temp_data, q);
+
+  } catch (sycl::exception const &e) {
+    std::cerr << "Caught a synchronous SYCL exception: " << e.what()
+              << std::endl;
+    std::terminate();
+  }
+}
+
+/////// HELPER FUNCTIONS ///////
+
+// provides a linear offset in the input array
+template <int n>
+int Coordinates(int iteration, int i) {
+  return iteration * n + i;
+}
+
+// Reference Fourier transform
+template <int lognr_points>
+void FourierTransformGold(ac_complex<double> *data, bool inverse) {
+  constexpr int kNrPoints = 1 << lognr_points;
+
+  // The inverse requires swapping the real and imaginary component
+  if (inverse) {
+    for (int i = 0; i < kNrPoints; i++) {
+      double tmp = data[i].r();
+      data[i].r() = data[i].i();
+      data[i].i() = tmp;
+    }
+  }
+
+  // Do a FT recursively
+  FourierStage<lognr_points>(data);
+
+  // The inverse requires swapping the real and imaginary component
+  if (inverse) {
+    for (int i = 0; i < kNrPoints; i++) {
+      double tmp = data[i].r();
+      data[i].r() = data[i].i();
+      data[i].i() = tmp;
+    }
+  }
+}
+
+template <int lognr_points>
+void FourierStage(ac_complex<double> *data) {
+  if constexpr (lognr_points > 0) {
+    constexpr int kNrPoints = 1 << lognr_points;
+
+    ac_complex<double> *half1 = (ac_complex<double> *)malloc(
+        sizeof(ac_complex<double>) * kNrPoints / 2);
+    ac_complex<double> *half2 = (ac_complex<double> *)malloc(
+        sizeof(ac_complex<double>) * kNrPoints / 2);
+
+    if (half1 == nullptr || half2 == nullptr) {
+      std::cerr << "Failed to allocate memory in validation function."
+                << std::endl;
+      std::terminate();
+    }
+
+    for (int i = 0; i < kNrPoints / 2; i++) {
+      half1[i] = data[2 * i];
+      half2[i] = data[2 * i + 1];
+    }
+
+    FourierStage<lognr_points - 1>(half1);
+    FourierStage<lognr_points - 1>(half2);
+
+    for (int i = 0; i < kNrPoints / 2; i++) {
+      data[i].r() = half1[i].r() +
+                    cos(2 * M_PI * i / kNrPoints) * half2[i].r() +
+                    sin(2 * M_PI * i / kNrPoints) * half2[i].i();
+      data[i].i() = half1[i].i() -
+                    sin(2 * M_PI * i / kNrPoints) * half2[i].r() +
+                    cos(2 * M_PI * i / kNrPoints) * half2[i].i();
+      data[i + kNrPoints / 2].r() =
+          half1[i].r() - cos(2 * M_PI * i / kNrPoints) * half2[i].r() -
+          sin(2 * M_PI * i / kNrPoints) * half2[i].i();
+      data[i + kNrPoints / 2].i() =
+          half1[i].i() + sin(2 * M_PI * i / kNrPoints) * half2[i].r() -
+          cos(2 * M_PI * i / kNrPoints) * half2[i].i();
+    }
+
+    free(half1);
+    free(half2);
+  } else {
```

Comment:
Why the "else { return }?    This is the default behavior from the function, right?

Reply:
Yes, having an assert here makes much more sense